feat(core): `SimpleChange` interface now accepts `T`

alan-agius4 opened the pull request:
…iousValue` and `currentValue` type

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
SimpleChange now supports generics

Issue Number: N/A


## What is the new behavior?
```
ngOnChanges(changes: {options?: SimpleChange<ComponentOptions>}) {
```

## Does this PR introduce a breaking change?
```
[X] Yes
[] No
```

This
<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information
Angular now requires TypeScript  `2.3.x` or greater as it has uses default generics

alan-agius4 committed:
feat(core): `SimpleChange` interface now accepts `T`

alan-agius4 committed:
test(forms): fix directives tests

alan-agius4 committed:
style(forms): fix directives formatting

Toxicable commented on packages/forms/test/directives_spec.ts:
```diff
@@ -603,15 +603,15 @@ export function main() {
            expect(ngModel.control.disabled).toEqual(false);
 
-           ngModel.ngOnChanges({isDisabled: new SimpleChange('', 0, false)});
+           ngModel.ngOnChanges({isDisabled: new SimpleChange<string|number>('', 0, false)});
```
Does the TS compiler throw an error here without the generic? if so then this would be considered a breaking change

alan-agius4 replied:
Isn't this internal? That said, I noticed it's a breaking change as default generics are available in TS 2.3.x>

Toxicable replied:
While the TS requirement itself is indeed a breaking change, other components of Angular are likely to also require it by v5 so you should be in the clear for that.
As for the other part, SimpleChange is a public API therefore if it's a breaking change for you to use it this way here then it will be for anyone else using it the same way

alan-agius4 replied:
Marked it, that said, AFAIK only the interface `SimpleChange` is public facing and not the class `SimpleChange` which are different.

Toxicable commented on packages/core/src/metadata/lifecycle_hooks.ts:
```diff
@@ -14,7 +14,7 @@ import {SimpleChange} from '../change_detection/change_detection_util';
  * @stable
  */
-export interface SimpleChanges { [propName: string]: SimpleChange; }
+export interface SimpleChanges { [propName: string]: SimpleChange<any>; }
```
Why `any` here?
You could make SimpleChanges generic then use a mapped types on simples changes to do a type lookup

alan-agius4 replied:
Yeah good spot, did it.

Commit:
feat(core): `SimpleChanges` now accepts `T` as argument

BREAKING CHANGE: Angular now requires TypeScript  `2.3.x` or greater as it has uses default generics

alan-agius4 commented:
Had to re-create the PR because of the CLA, as my email is different and couldn't link it with my CLA email

New PR https://github.com/angular/angular/pull/18444
